packages/builder/src/components/automation/AutomationBuilder/FlowChart/FlowChart.svelte

```diff
@@ -23,49 +31,226 @@
+  import TestDataModal from "./TestDataModal.svelte"
+  import NodeWrapper from "./NodeWrapper.svelte"
+  import CustomEdge from "./CustomEdge.svelte"
+  import BranchNodeWrapper from "./BranchNodeWrapper.svelte"
+  import AnchorNode from "./AnchorNode.svelte"
 
-  export let automation
+  import {
+    SvelteFlow,
+    Background,
+    BackgroundVariant,
+    MiniMap,
+    useSvelteFlow,
+    type Node as FlowNode,
+    type Edge as FlowEdge,
+    type NodeTypes,
+    type EdgeTypes,
+  } from "@xyflow/svelte"
+  import "@xyflow/svelte/dist/style.css"
+  import FlowControls from "./Controls.svelte"
+
+  export let automation: UIAutomation
 
   const memoAutomation = memo(automation)
 
-  let testDataModal
-  let confirmDeleteDialog
-  let blockRefs = {}
-  let treeEle
-  let draggable
-  let prodErrors
-  let viewMode = ViewMode.EDITOR
+  const nodeTypes: NodeTypes = {
+    "step-node": NodeWrapper as any,
+    "branch-node": BranchNodeWrapper as any,
+    "anchor-node": AnchorNode as any,
+  }
+  const edgeTypes: EdgeTypes = {
+    "add-item": CustomEdge as any,
+  }
 
+  let testDataModal: Modal
+  let confirmDeleteDialog
+  let blockRefs: Record<string, any> = {}
+  let prodErrors: number
+  let paneEl: HTMLDivElement | null = null
   let changingStatus = false
 
+  let initialViewportApplied = false
+  let preserveViewport = false
+  let layoutDirection: LayoutDirection = automation.layoutDirection || "TB"
+
+  let nodes = writable<FlowNode[]>([])
+  let edges = writable<FlowEdge[]>([])
+
+  const { getViewport, setViewport, fitView } = useSvelteFlow()
+
+  // DnD helper and context stores
+  const dnd = createFlowChartDnD({
+    getViewport,
+    setViewport,
+    moveBlock: (sourcePath, destPath, automationData) =>
+      automationStore.actions.moveBlock(sourcePath, destPath, automationData),
+    getSelectedAutomation: () => get(selectedAutomation),
+  })
+  const { view, viewPos, contentPos } = dnd
+  setContext("draggableView", view)
+  setContext("viewPos", viewPos)
+  setContext("contentPos", contentPos)
+
+  $: updateGraph(blocks as any, viewMode, layoutDirection)
+
   $: $automationStore.showTestModal === true && testDataModal.show()
 
   $: displayToggleValue =
     automation.publishStatus.state === PublishResourceState.PUBLISHED
 
   // Memo auto - selectedAutomation
-  $: memoAutomation.set(automation)
+  $: memoAutomation.set($selectedAutomation.data || automation)
 
   // Parse the automation tree state
-  $: refresh($memoAutomation)
-
-  $: blocks = getBlocksHelper($memoAutomation, viewMode).filter(
-    x => x.stepId !== ActionStepID.LOOP
+  $: $selectedAutomation.blockRefs && refresh()
+  $: blocks = getBlocksHelper(
+    $selectedAutomation.data || $memoAutomation,
+    viewMode
   )
+    .filter(x => x.stepId !== ActionStepID.LOOP)
+    .map((block, idx) => ({ ...block, __top: idx }))
+
+  $: viewMode = ViewMode.EDITOR
+
+  const updateGraph = async (
+    blocks: any,
+    currentViewMode: any,
```

Comment:
Don't use any type, but also could use object parameters to better expose the inputs.

```ts
{ blocks, currentViewMode, direction } : { blocks: Type,
    currentViewMode: Type, direction: LayoutDirection }
```